File: frontend/src/types/graph.ts
/**
 * types/graph.ts
 *
 * Why this file exists:
 *   Mirrors backend/app/schemas/graph.py exactly. This is the frontend
 *   half of the Universal Graph contract - keeping both sides in sync
 *   manually (rather than auto-generating) is fine at this project size,
 *   but if the backend schema changes, this file must change with it.
 *
 * How it connects:
 *   Used by api/client.ts (typed fetch responses) and every component
 *   that renders graph data (GraphCanvas, LayerPropertiesPanel, etc).
 */

export type Framework = "pytorch" | "tensorflow" | "jax" | "unknown";
export type Confidence = "traced" | "static" | "partial";
export type GroupType = "conv_block" | "residual_block" | "stage";

export interface GraphGroup {
  id: string;
  label: string;
  type: GroupType;
  member_node_ids: string[];
  parent_group_id: string | null;
  repeat_count: number;
}

export interface GraphNode {
  id: string;
  type: string;
  label: string;
  input_shape: number[] | null;
  output_shape: number[] | null;
  params: number;
  group_id: string | null;
}

export interface GraphEdge {
  source: string;
  target: string;
  is_skip_connection: boolean;
}

export interface GraphMeta {
  framework: Framework;
  confidence: Confidence;
  total_params: number;
  total_layers: number;
  flops: number | null;
  warnings: string[];
}

export interface UniversalGraph {
  job_id: string;
  model_name: string;
  meta: GraphMeta;
  nodes: GraphNode[];
  edges: GraphEdge[];
  groups: GraphGroup[];
}

export interface UploadResponse {
  job_id: string;
  filename: string;
  status: string;
}

File: neuralviz-cli/neuralviz/_vendored/schemas/graph.py
# VENDORED COPY — synced from backend/app/schemas/graph.py
# This file is part of the neuralviz CLI package's vendored engine layer.
# If you update backend/app/schemas/graph.py, manually sync changes here.
# See neuralviz-cli/README.md for the sync procedure.

"""
schemas/graph.py

Defines the Universal Graph JSON contract as Pydantic models. This is
the single, fixed data shape that every framework parser (PyTorch,
TensorFlow, JAX, custom) must produce, and that the frontend consumes.
"""

from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class Framework(str, Enum):
    PYTORCH = "pytorch"
    TENSORFLOW = "tensorflow"
    JAX = "jax"
    ONNX = "onnx"
    UNKNOWN = "unknown"


class Confidence(str, Enum):
    TRACED = "traced"           # produced by executing the model (torch.fx)
    STATIC = "static"           # produced by reading source only (AST)
    PARTIAL = "partial"         # some nodes could not be resolved


class GraphNode(BaseModel):
    id: str = Field(..., description="Unique node identifier, e.g. 'node_3'")
    type: str = Field(..., description="Operation/layer type, e.g. 'Conv2d'")
    label: str = Field(..., description="Human-readable name, e.g. 'conv1'")
    input_shape: Optional[list[int]] = None
    output_shape: Optional[list[int]] = None
    params: int = 0
    flops: Optional[int] = Field(default=None, description="FLOP count for this node/layer")
    line_number: Optional[int] = Field(default=None, description="Source line number")
    group_id: Optional[str] = Field(default=None, description="Set by grouping engine")


class GraphEdge(BaseModel):
    source: str
    target: str
    is_skip_connection: bool = False


class GroupType(str, Enum):
    CONV_BLOCK = "conv_block"
    RESIDUAL_BLOCK = "residual_block"
    STAGE = "stage"


class GraphGroup(BaseModel):
    id: str = Field(..., description="Unique group identifier, e.g. 'group_3'")
    label: str = Field(..., description="Display label, e.g. 'ConvBlock'")
    type: GroupType
    member_node_ids: list[str] = Field(default_factory=list)
    parent_group_id: Optional[str] = Field(default=None)
    repeat_count: int = Field(default=1)


class GraphMeta(BaseModel):
    framework: Framework
    confidence: Confidence
    total_params: int = 0
    total_layers: int = 0
    flops: Optional[int] = Field(default=None)
    warnings: list[str] = Field(default_factory=list)


class UniversalGraph(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    job_id: str
    model_name: str
    meta: GraphMeta
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    groups: list[GraphGroup] = Field(default_factory=list)


class UploadResponse(BaseModel):
    job_id: str
    filename: str
    status: str = "uploaded"


class SourceResponse(BaseModel):
    job_id: str
    filename: str
    code: str


class JobStatusResponse(BaseModel):
    job_id: str
    status: str
    detail: Optional[str] = None


class ProjectMetadata(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    detected_framework: str
    primary_model_file: str
    files_scanned: int
    model_candidates: list[str]
    dependencies: list[str]
    warnings: list[str]


class ProjectAnalysisResponse(BaseModel):
    metadata: ProjectMetadata
    graph: UniversalGraph
